### ISSUE TITLE
GPU Optimization

### ISSUE THREAD

**Issue by ashissamal** [Ki6an/fastT5#34]
Thanks for sharing the repo . It is really helpful.

I'm exploring ways to do the optimization on GPU. I know its not presently supported. Could you share some approach or references to implement the optimization on GPU(Nvidia)

**Comment by Ki6an** [comment]
for GPU you can use the [`onnxruntime-gpu` ](https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#contents)library, but it does not support quantization. so you won't have the advantage of reduced model size during inference.  

[here's](https://github.com/microsoft/onnxruntime/blob/dfe42e185c6c6de68177db8ecf307645ce831aec/onnxruntime/python/tools/transformers/notebooks/PyTorch_Bert-Squad_OnnxRuntime_GPU.ipynb) an example implementation of this library for BERT, you can follow this guide and make suitable changes for T5. In addition to this you also need to implement[ `iobinding` ](http://www.xavierdupre.fr/app/onnxruntime/helpsphinx/api_summary.html#id4). 
 I tried without iobinding but wasn't able to get any advantages over pytorch.

**Comment by sam-writer** [comment]
I would also check out this recent demo that NVIDIA did of TensorRT, which involves converting to ONNX as an intermediate step. They run the tests on GPU https://github.com/NVIDIA/TensorRT/blob/main/demo/HuggingFace/README.md

**Comment by pommedeterresautee** [comment]
ONNX Runtime supports GPU quantization through the TensorRT provider (now embedded by default in the GPU version of the pipy lib, no need for a custom compilation). However it only supports PTQ, meaning there is a 2-3 point accuracy cost (VS QAT or dynamic quantization which are usually close to non quantized accuracy). Quantization bring a X2 speedup, that you can add to a 1.3 speedup when you switch from ORT to TRT, so quite significant on base / large models (not yet benchmarked on distilled models)

Hopefully, QAT is also doable but requires some work per model. You can see some here https://github.com/ELS-RD/transformer-deploy/pull/29, for now only for Albert, Electra, Bert, Roberta and Distilbert. I will probably add support for Deberta V1 and V2, T5 and Bart as a next step.

**Comment by nbravulapalli** [comment]
I really appreciate the functionality that the fastT5 library offers!

Like the original poster, I am looking to leverage the speedup from both ONNX Runtime and quantization that fastT5 offers, and deploy this on a Nvidia GPU. Do you have any pointers on how to accomplish this with a t5-large model? 

@Ki6an or @sam-writer
Are there plans to add GPU support for this library?

Thanks!

**Comment by sam-writer** [comment]
@nbravulapalli yes, there are plans to make running on GPU as easy as running on CPU is currently. However, if you need to run on GPU now, your best bet is probably to follow [this notebook](https://github.com/NVIDIA/TensorRT/blob/main/demo/HuggingFace/notebooks/t5.ipynb) to convert the model to TensorRT format, which runs on GPU faster than quantized ONNX t5 runs on CPU.

**Comment by sam-writer** [comment]
Here is an example from a branch of the ONNX library that demonstrates using `io-binding` as well as other tricks needed to run on GPU [link](https://github.com/microsoft/onnxruntime/blob/tlwu%2Fbenchmark_t5/onnxruntime/python/tools/transformers/t5/t5_helper.py)

**Comment by GenVr** [comment]
@sam-writer So currently to get a performance optimization for inference time on T5 on GPU, do you recommend [this ](https://github.com/microsoft/onnxruntime/blob/tlwu/benchmark_t5/onnxruntime/python/tools/transformers/t5/benchmark_t5.py#L290)code?

> Here is an example from a branch of the ONNX library that demonstrates using `io-binding` as well as other tricks needed to run on GPU [link](https://github.com/microsoft/onnxruntime/blob/tlwu%2Fbenchmark_t5/onnxruntime/python/tools/transformers/t5/t5_helper.py)

Is there an example-code of real use?

I would like to improve the GPU inference time with a T5-base with `max_length=1024`.
